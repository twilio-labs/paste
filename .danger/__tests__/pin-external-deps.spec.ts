import pinExternalDeps, {
  getUnPinnedExternalDepsFromPackageJSONFiles,
  getUnpinnedExternalDeps,
} from '../pin-external-deps';

declare const global: any;

describe('pin external dependencies check', () => {
  describe('getUnpinnedExternalDeps()', () => {
    it('should return the names of external deps that are not pinned', () => {
      expect(getUnpinnedExternalDeps({lodash: '^4.17.21'})).toEqual(['lodash']);
      expect(getUnpinnedExternalDeps({csstype: '3.0.11', lodash: '^4.17.21'})).toEqual(['lodash']);
      expect(getUnpinnedExternalDeps({csstype: '^3.0.11', lodash: '^4.17.21'})).toEqual(['csstype', 'lodash']);
      expect(getUnpinnedExternalDeps({csstype: '3.0.11'})).toEqual([]);
    });
  });
  describe('getUnPinnedExternalDepsFromPackageJSONFiles()', () => {
    it('should return an array of all the unpinned external deps across every provided package.json file', () => {
      expect(getUnPinnedExternalDepsFromPackageJSONFiles(['./.danger/__fixtures__/mock_package/package.json'])).toEqual(
        ['lodash']
      );

      expect(
        getUnPinnedExternalDepsFromPackageJSONFiles(['./.danger/__fixtures__/mock_package_2/package.json'])
      ).toEqual(['csstype', 'lodash']);

      expect(
        getUnPinnedExternalDepsFromPackageJSONFiles(['./.danger/__fixtures__/mock_package_3/package.json'])
      ).toEqual([]);

      expect(
        getUnPinnedExternalDepsFromPackageJSONFiles([
          './.danger/__fixtures__/mock_package/package.json',
          './.danger/__fixtures__/mock_package_3/package.json',
        ])
      ).toEqual(['lodash']);

      expect(
        getUnPinnedExternalDepsFromPackageJSONFiles([
          './.danger/__fixtures__/mock_package/package.json',
          './.danger/__fixtures__/mock_package_2/package.json',
          './.danger/__fixtures__/mock_package_3/package.json',
        ])
      ).toEqual(['lodash', 'csstype', 'lodash']);
    });
  });

  describe('pinExternalDeps()', () => {
    beforeEach(() => {
      global.warn = jest.fn();
      global.message = jest.fn();
      global.fail = jest.fn();
      global.markdown = jest.fn();
    });

    afterEach(() => {
      global.warn = undefined;
      global.message = undefined;
      global.fail = undefined;
      global.markdown = undefined;
      global.danger = undefined;
    });

    it('should not fail when there are not unpinned external deps', () => {
      global.danger = {
        git: {
          modified_files: ['./.danger/__fixtures__/mock_package_3/package.json'],
          created_files: [],
        },
      };
      pinExternalDeps();
      expect(global.fail).not.toHaveBeenCalled();
    });

    it('should fail when there are unpinned external deps', () => {
      global.danger = {
        git: {
          modified_files: ['./.danger/__fixtures__/mock_package_2/package.json'],
          created_files: [],
        },
      };
      pinExternalDeps();
      expect(global.fail).toHaveBeenCalled();
    });

    it('should fail even when there are mixed pinned and unpinned external deps', () => {
      global.danger = {
        git: {
          modified_files: ['./.danger/__fixtures__/mock_package/package.json'],
          created_files: [],
        },
      };
      pinExternalDeps();
      expect(global.fail).toHaveBeenCalled();
    });
  });
});
