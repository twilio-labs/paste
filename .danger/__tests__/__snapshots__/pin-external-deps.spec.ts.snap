// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pin external dependencies check getUnPinnedExternalDepsFromPackageJSONFiles() should return an array of all the unpinned external deps across every provided package.json file 1`] = `
Array [
  Object {
    "dep": "lodash",
    "name": "mock-package",
  },
]
`;

exports[`pin external dependencies check getUnPinnedExternalDepsFromPackageJSONFiles() should return an array of all the unpinned external deps across every provided package.json file 2`] = `
Array [
  Object {
    "dep": "csstype",
    "name": "mock-package-2",
  },
  Object {
    "dep": "lodash",
    "name": "mock-package-2",
  },
]
`;

exports[`pin external dependencies check getUnPinnedExternalDepsFromPackageJSONFiles() should return an array of all the unpinned external deps across every provided package.json file 3`] = `Array []`;

exports[`pin external dependencies check getUnPinnedExternalDepsFromPackageJSONFiles() should return an array of all the unpinned external deps across every provided package.json file 4`] = `
Array [
  Object {
    "dep": "lodash",
    "name": "mock-package",
  },
]
`;

exports[`pin external dependencies check getUnPinnedExternalDepsFromPackageJSONFiles() should return an array of all the unpinned external deps across every provided package.json file 5`] = `
Array [
  Object {
    "dep": "lodash",
    "name": "mock-package",
  },
  Object {
    "dep": "csstype",
    "name": "mock-package-2",
  },
  Object {
    "dep": "lodash",
    "name": "mock-package-2",
  },
]
`;
